lang/step.md

```diff
@@ -533,18 +545,28 @@ impl<M: Memory> Machine<M> {
 
         // Evaluate all arguments.
         let arguments = arguments.try_map(|arg| self.eval_value(arg))?;
-        self.eval_intrinsic(intrinsic, arguments, ret_place)?;
 
-        if let Some(next_block) = next_block {
-            self.mutate_cur_frame(|frame| {
-                frame.jump_to_block(next_block);
-            });
-        } else {
-            throw_ub!("Return from an intrinsic where caller did not specify next block");
+        match self.eval_intrinsic(intrinsic, arguments, ret_place)? {
+            ExecutionStatus::Finished => {
+                if let Some(next_block) = next_block {
+                    self.mutate_cur_frame(|frame| {
+                        frame.jump_to_block(next_block);
+                    });
+                } else {
+                    throw_ub!("Return from an intrinsic where caller did not specify next block");
+                }
+            },
+            ExecutionStatus::Blocked => {
+                // In the future once this thread is `Enabled` again, we retry this intrinsic.
+                
+                // If an execution was blocked, the thread can't be enabled.
+                // This is a sanity check that all intrinsics are implemented correctly.
+                assert!(self.cur_thread().state != ThreadState::Enabled);
+            },
```

Comment:
If I understand the code here correctly, then once the thread gets unblocked, it will run *the same intrinsic again*? That seems wrong, unblocking should make it *return from the intrinsic*.

So, why do you even need `ExecutionStatus`? Can't we do `jump_to_block` in both cases? Just in one case the thread will be blocked so it will take a while until the next statement is executed.

Reply:
True. I thought this will be useful for locks, because I can simply wake all threads and put them back to sleep. But it would be easier to just wake an arbitrary.

Reply:
If we go down that route, we don't need a 'blocked' state at all, we can just make Join spin until the target thread is finished.

Properly blocking threads is more realistic though so let's go for that.